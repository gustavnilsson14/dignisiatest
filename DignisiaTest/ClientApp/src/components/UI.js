import React, { Component } from 'react';
//﻿import { InputGroup, Dropdown, FormControl,DropdownToggle,DropdownMenu,DropdownItem } from 'reactstrap';
import { Button, Form, FormGroup, Label, Input, FormText, Dropdown, DropdownToggle, DropdownMenu, DropdownItem, Container, Row, Col } from 'reactstrap';
import {MockData} from '../mock/MockData.js'
const ReactHighcharts = require('react-highcharts');

class UI extends React.Component {
  constructor(props) {
    super(props);
    this.state = { segment: null, segmentId: 1, caseAgeRangeEnd: 1 };
  }
	onFormChange = (newState) => {
		this.setState(newState);
	}
	render() {
		return (
			<div>
				<UIControls startDate="2020-01-01" endDate="2021-02-01" onFormChange={this.onFormChange.bind(this)} />
				<Container>
				  <Row>
				    <Col>
							<DebtSizeWidget key={this.state.segmentId + this.state.caseAgeRangeEnd} {...this.state}/>
						</Col>
				    <Col>
							<CaseOverviewWidget key={this.state.segmentId + this.state.caseAgeRangeEnd} {...this.state}/>
						</Col>
				  </Row>
				</Container>
			</div>
		);
	}
}
class UIControls extends React.Component {
  //const [dropdownOpen, setDropdownOpen] = useState(false);

  constructor(props) {
    super(props);
    this.state = { segment: null, segmentId: 1, caseAgeRangeEnd: 0, loading: true };
  }
  componentDidMount = () => {
    this.getSegments();
  }
  toggle = () => {
		this.setState({isOpen: !this.isOpen()});
	}
	isOpen = () => {
		if (!this.state){
			return false;
		}
		return this.state.isOpen;
	}
	getSegmentOptionsRecursive = (segment, parentSegment = null) => {
		if (segment == null)
			return null;
		var result = [this.getSegmentOption(segment, parentSegment)];
		segment.children.forEach((child, i) => {
			result = [result, ...this.getSegmentOptionsRecursive(child, segment)];
		});
		return result;
	}
	getSegmentOption = (segment, parentSegment) => {
		var segmentDisplayName = segment.segmentName;
		if (parentSegment != null)
			segmentDisplayName = `${parentSegment.segmentName} - ${segment.segmentName}`;
		return(
			<option key={segment.segmentId} value={segment.segmentId}>{segmentDisplayName}</option>
		);
	}
	getCaseAgeRange = () => {
		var difference = this.getMonthDifference(new Date(this.props.startDate), new Date(this.props.endDate))
		return difference;
	}
	getCaseAgeRangeEndDate = () => {
		var endDate = new Date(this.props.startDate);
		endDate.setMonth(endDate.getMonth() + this.state.caseAgeRangeEnd);
		return endDate.toISOString().split('T')[0];
	}
	onCaseAgeRangeChange = (event) => {
		var newState = {caseAgeRangeEnd: parseInt(event.target.value)};
		this.props.onFormChange(newState);
		this.setState(newState);
	}
	onSegmentSelectChange = (event) => {
		var newState = {segmentId: event.target.value};
		this.props.onFormChange(newState);
		this.setState(newState);
	}
	getMonthDifference = (d1, d2) => {
		return d2.getMonth() - d1.getMonth() +
   (12 * (d2.getFullYear() - d1.getFullYear()))
	}
	async getSegments(){
	    const response = await fetch('segments');
	    const data = await response.json();
			this.setState({ segment: data });
	}
	render() {
		return (
			<Form>
	      <FormGroup>
	        <Label for="segmentSelect">Segment</Label>
	        <Input value={this.state.segmentId} onChange={this.onSegmentSelectChange} type="select" name="segmentSelect" id="segmentSelect">
						{this.getSegmentOptionsRecursive(this.state.segment)}
	        </Input>
	      </FormGroup>
		      <FormGroup>
		        <Label for="caseAgeRange">Case Age</Label>
						<span style={{float:"right"}}>From {this.props.startDate} to {this.getCaseAgeRangeEndDate()}</span>
						<Input onChange={this.onCaseAgeRangeChange} type="range" value={this.state.caseAgeRangeEnd} min="0" max={this.getCaseAgeRange()} step="1" name="caseAgeRange" id="caseAgeRange"/>
						<span>{this.props.startDate}</span>
						<span style={{float:"right"}}>{this.props.endDate}</span>
		      </FormGroup>
	    </Form>
		);
	}
}
class DebtSizeWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = { chartConfig: [], loading: true };
  }
	componentDidMount = () => {
		this.getChartConfig();
	}
	async getChartConfig(){
		var api = `chart-data/1?segmentId=${this.props.segmentId}&caseAge=${this.props.caseAgeRangeEnd}`;
		const response = await fetch(api);
    const data = await response.json();
		this.setState({ chartConfig: MockData.debtWidgetMockData(data) });
	}
	render() {
		return (
			<ReactHighcharts config={this.state.chartConfig}></ReactHighcharts>
		);
	}
}
class CaseOverviewWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = { chartConfig: [], loading: true };
  }
	componentDidMount = () => {
		this.getChartConfig();
	}
	async getChartConfig(){
		var api = `chart-data/2?segmentId=${this.props.segmentId}&caseAge=${this.props.caseAgeRangeEnd}`;
    const response = await fetch(api);
    const data = await response.json();
		this.setState({ chartConfig: MockData.caseOverviewMockData(data) });
	}
	render() {
		return (
			<ReactHighcharts config={this.state.chartConfig}></ReactHighcharts>
		);
	}
}

export { UI, UIControls };
