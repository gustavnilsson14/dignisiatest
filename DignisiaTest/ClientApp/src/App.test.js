// Link.react.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import "regenerator-runtime/runtime";
import { UI, UIControls } from './components/UI.js';

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <UIControls startDate="2020-01-01" endDate="2021-02-01" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
