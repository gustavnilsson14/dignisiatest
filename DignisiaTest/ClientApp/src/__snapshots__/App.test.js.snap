// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<form
  className=""
>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="segmentSelect"
    >
      Segment
    </label>
    <select
      className="form-control"
      id="segmentSelect"
      name="segmentSelect"
      onChange={[Function]}
      value={1}
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
      htmlFor="caseAgeRange"
    >
      Case Age
    </label>
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      From 
      2020-01-01
       to 
      2020-01-01
    </span>
    <input
      className="form-control"
      id="caseAgeRange"
      max={13}
      min="0"
      name="caseAgeRange"
      onChange={[Function]}
      step="1"
      type="range"
      value={0}
    />
    <span>
      2020-01-01
    </span>
    <span
      style={
        Object {
          "float": "right",
        }
      }
    >
      2021-02-01
    </span>
  </div>
</form>
`;
